src/dim_helpers/ConvDims.jl

```diff
@@ -6,34 +6,54 @@ export ConvDims
 Type system-level information about convolution dimensions. Critical for things like
 `im2col!()` to generate efficient code, and helpful to reduce the number of kwargs
 getting passed around.
-
-We don't want to specialize on things like image size/channel count, so we generally
-store those as fields, just for convenience, and to allow for non-breaking changes when
-we decide we _do_ want to specialize on those values.  We always want to specialize on
-things like stride, padding, dilation, and kernel flipping though.
 """
-abstract type ConvDims{N, S, P, D, F} end
+abstract type ConvDims{N} end
+
+@inline spatial_dims(::ConvDims{N}) where N = N
+@inline input_size(c::ConvDims) = c.input_size
+@inline kernel_size(c::ConvDims) = c.kernel_size
+
+@inline stride(c::ConvDims) = c.stride
+@inline padding(c::ConvDims) = c.padding
+@inline dilation(c::ConvDims) = c.dilation
```

Comment:
Doesn't seem to affect much. Maybe slightly slower or maybe just noise.
But as for the default values, we have them in the constructors for other `...Dims`, so no sure they are needed here.
Otherwise we can add following defaults:

```julia
stride(c::ConvDims{N}) = NTuple{N, Int}(repeat(1, N))
padding(c::ConvDims{N}) = NTuple{N, Int}(repeat(0, N))
dilation(c::ConvDims{N}) = NTuple{N, Int}(repeat(1, N))
```

Reply:
Ok, let's not worry about defaults then.